src/Controller/IdentityController.php

```diff
@@ -0,0 +1,153 @@
+ */
+
+namespace Surfnet\GsspBundle\Controller;
+
+use RuntimeException;
+use SAML2_Const;
+use SAML2_Response;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Surfnet\GsspBundle\Saml\ResponseBuilderFactory;
+use Surfnet\GsspBundle\Saml\ResponseContextInterface;
+use Surfnet\GsspBundle\Saml\StateHandler;
+use Surfnet\GsspBundle\Service\ConfigurationContainer;
+use Surfnet\GsspBundle\Service\ResponseServiceInterface;
+use Surfnet\SamlBundle\Http\RedirectBinding;
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @Route(service="surfnet_gssp.saml.identity_controller")
+ */
+final class IdentityController extends Controller
+{
+    private $httpBinding;
+    private $registrationRoute;
+    private $stateHandler;
+    private $responseService;
+    private $responseContext;
+    private $responseBuilderFactory;
+
+    public function __construct(
+        RedirectBinding $httpBinding,
+        ConfigurationContainer $configuration,
+        StateHandler $stateHandler,
+        ResponseServiceInterface $responseService,
+        ResponseContextInterface $responseContext,
+        ResponseBuilderFactory $responseBuilderFactory
+    ) {
+        $this->registrationRoute = $configuration;
+        $this->stateHandler = $stateHandler;
+        $this->responseService = $responseService;
+        $this->responseContext = $responseContext;
+        $this->httpBinding = $httpBinding;
+        $this->responseBuilderFactory = $responseBuilderFactory;
+    }
+
+    /**
+     * @Route("/saml/sso", name="gssp_saml_sso", methods={"GET"})
+     * @throws \InvalidArgumentException
+     */
+    public function ssoAction(Request $request)
+    {
+        // If we already have a request, we clear the current state.
+        if ($this->responseContext->hasRequest()) {
+            $this->stateHandler->invalidate();
+        }
+
+        try {
+            $originalRequest = $this->httpBinding->receiveSignedAuthnRequestFrom($request);
+        } catch (RuntimeException $e) {
+            return $this->render('@SurfnetGssp/StepupGssp/unrecoverableError.html.twig', [
+                'message' => $e->getMessage(),
+            ], new Response(null, Response::HTTP_NOT_ACCEPTABLE));
+        }
+        $this->stateHandler
+            ->setRequestId($originalRequest->getRequestId())
+            ->setRequestServiceProvider($originalRequest->getServiceProvider())
+            ->setRelayState($request->get(AuthnRequest::PARAMETER_RELAY_STATE, ''));
+        return new RedirectResponse($this->generateUrl($this->registrationRoute->getRegistrationRoute()));
+    }
+
+    /**
+     * @Route("/saml/sso_return", name="gssp_saml_sso_return", methods={"POST", "GET"})
+     * @throws \Exception
+     */
+    public function ssoReturnAction()
```

Comment:
Consider describing the intent of this controller action. This comment goes for the other actions as well.